biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts

```diff
@@ -1,20 +1,269 @@
+    buildings: Building[],
+    smallBatteries: BioBattery[],
+    largeBatteries: BioBattery[],
+    solarPanels: SolarPanel[],
+    shortestDistances: {[source: string]: { [node: string]: Path}}
+  ): SupplyingPath {
+    // Assuming that the houses asking for power will not have power in them.
+    // Do not consider building with full power capacity
+    buildings = buildings.filter((building) => {
+      return building.getEnergyInJoules() === building.MinCapacity;
+    });
+
+    // Filter the batteries and removes the ones which do not have power in them
+    // Do not include batteries which are empty
+    smallBatteries = smallBatteries.filter((battery) => !battery.isEmpty());
+    largeBatteries = largeBatteries.filter((battery) => !battery.isEmpty());
+    
+    // Filter the solar panels and remove the ones with the minimum energy or empty
+    solarPanels = solarPanels.filter((solarPanel) => !solarPanel.isEmpty());
+
+    // Create an array of the possible energy givers
+    const allEnergyProviders = [
+      ...smallBatteries,
+      ...solarPanels,
+      ...largeBatteries,
+    ];
+
+    return this.determineSupplyingPath(buildings, allEnergyProviders, shortestDistances);
+  }
+
+  /**
+   * This method determines which grid item can supply energy to the grid item without it.
+   * It determines this by considering the needs of the grid item requesting the enrgy and 
+   * minimizing the distance between the supplying grid items and the receiver. It returns a key pair of receiver to supplier
+   * @param recievingAgents holds a list of grid items (buildings or batteries but not both) which are requesting for energy
+   * @param supplyingAgents holds a list of grid items (@class BioBattery or @class SolarPanel) 
+   * which can supply energy to @param recievingAgents
+   * @param shortestDistances holds an object of key, value pair of vertex -> adj vertices with their shortest distance to the key vertex
+   * @returns @interface SupplyingPath which holds a key value pair of a gridItem requesting mapping to the one which can supplying
+   */
+  private determineSupplyingPath(
+    recievingAgents: RecievingAgents,
+    supplyingAgents: SupplyingAgents[],
+    shortestDistances: ShortestDistances
+  ): SupplyingPath {
+    // TODO: advancement for later implementations: assign the recievingAgent itself not the names
+    // Create an object of buildings with the energyProviders which supplied
+    let supplyToSupplyFromAgents: SupplyingPath = {};
+    // Look at each gridItem requesting for energy individually and keep track of the which grid item
+    //  supplied it energy so that it can not be requested energy when it doesn't have it
+    for (const recievingAgent of recievingAgents) {
+      // get the energy which is being requested.
+      // For now implement all or nothing. If battery doesn't have all the energy required, ignore it
+      const energyReq = recievingAgent.MaxCapacity - recievingAgent.getEnergyInJoules();
```

Comment:
instead of calling the variable directly, we should be calling a getter method (should be `receivingAgent.getMaxCapacity()`

Reply:
I understand that this works as is, but prefer using getter methods and staying consistent